### PR TITLE
Added auto screenshot module

### PR DESCRIPTION
**Pull request by sphvn**
@kaytotes Pulled the idea for this one out of the project board.

### CODE REVIEW

File: modules/misc/autoscreenshot.lua [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    @ param Frame $self The Frame that is handling the event 
+    @ param string $event The WoW API Event that has been triggered
+    @ param arg $... The arguments of the Event
+    @ return void
+]]
+local function HandleEvents (self, event, ...)
+    if (event == 'ACHIEVEMENT_EARNED' and PrimaryDB.autoScreenshot) then
+        Screenshot();
+    end
+    -- Should we handle SCREENSHOT_SUCCEEDED / SCREENSHOT_FAILED, probably not our business.
```
Comment (sphvn) [inline modules/misc/autoscreenshot.lua]: Figured handling `SCREENSHOT_SUCCEEDED` or `SCREENSHOT_FAILED` was a bit over-reaching for what this should be doing, figured we should just try and leave it at that, maybe print a message to chat afterwards?

**Comment by sphvn** [comment]
Let me know if there is anything you want different, happy to follow suite.

**Comment by sphvn** [comment]
Yeah it does display the screenshot message, you can't bypass that.  If multiple achievements are earned in a row sadly then the message will show up in the secondary and so forth screenshots but I think for the regular functionality this is probably a good thing anyway, it's pretty clear whats happening without having to print an extra message to chat

**Comment by sphvn** [comment]
And yeah totally the layout made it super easy to add this as a self contained thing, well done on that 👍

**Comment by sphvn** [comment]
Tested this generally with a new allied race and procuring the "Going Down" and the "Shave and a Haircut" achievements as it was fairly fast to replicate.

**Comment by sphvn** [comment]
Only thing I can think of is it might be worth a slight delay before the screenshot is taken, I am using is:toasts so that might just be for me.

wowscrnshot_021018_125424

**Comment by kaytotes** [comment]
> If multiple achievements are earned in a row sadly then the message will show up in the secondary and so forth screenshots

Hmm, this gets me thinking. Do you have an example output of when this executes? As in the screenshot it produces when a single achievement is earned. The achievement UI has a fade in transition until it's at full opacity as far as I'm aware so it might be best if this has something like a 1 or 2 second delay.

With regards to taking multiple screenshots in a row I know I'd personally find it annoying and I feel the text from the previous screenshot being in the current screenshot is a negative. I reckon the best way to handle this would be to roughly time how long it takes for the Blizzard screenshot text to disappear completely and limit the taking of screenshots to only be after that interval.

Probably the best way to do this would be to hook OnUpdate such as how the [new Performance Frame](https://github.com/kaytotes/ImprovedBlizzardUI/blob/feature/big-overhaul/modules/frames/performance.lua) handles it. In that case it has a delay of 2 seconds between ticks.

If you feel that would be out of the scope of the PR @sphvn just let me know and I can merge it in as is then do the final tweaks myself. Up to you.

**Comment by sphvn** [comment]
I'd say this covers most of the functionality, lets merge it and I'll work on a new PR with the extra tweaks, may as well merge often if it's not breaking things :)

**Comment by sphvn** [comment]
Also I can't think of any achievements that would be easy to earn in rapid succession, although copying the functionality, just make a macro to screenshot and click it twice, certainly isn't ideal.

**Comment by kaytotes** [comment]
I'd rather merge as a completed module rather than two seperate PR. I can't think of any easily repeatable ones either for testing that use case. I'd say it's best to put the delay in nonetheless for the sake of clean screenshots.

**Comment by sphvn** [comment]
Sorry I'm being slow with this, I will get to it :)

**Comment by sphvn** [comment]
> Edit: You posted as I was typing. Try it without ls:toasts to be sure. By the way, 4ms ping. Sweet jesus. 👍

Yeah it's certainly not bad 😄 , I really do like the display for the performance frame because I can see the differences playing on oceanic server sometimes it spikes to 5, when in battlegrounds and I get shifted to a US server I can see the total jump to 250ish, very clear indication of whats going on.

Just rebased this now and will finish it up this evening as I have some time.

> Probably the best way to do this would be to hook OnUpdate such as how the new Performance Frame handles it. In that case it has a delay of 2 seconds between ticks.

I quite like the way the Performance Frame is working I'll probably leverage this quite heavily.

Will turn off is:toasts and do some play testing also.

File: modules/misc/autoscreenshot.lua [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    @ param Frame $self The Frame that is handling the event 
+    @ param string $event The WoW API Event that has been triggered
+    @ param arg $... The arguments of the Event
+    @ return void
+]]
+local function HandleEvents (self, event, ...)
+    if (event == 'ACHIEVEMENT_EARNED' and PrimaryDB.autoScreenshot) then
+        AutoScreenshotFrame:UnregisterEvent('ACHIEVEMENT_EARNED');
+        C_Timer.After(1, Screenshot);
+        C_Timer.After(4, function() AutoScreenshotFrame:RegisterEvent('ACHIEVEMENT_EARNED') end);
```
Comment (sphvn) [inline modules/misc/autoscreenshot.lua]: @kaytotes How do you feel about debouncing the event this way rather than having to hit the onupdate every frame,

**Comment by sphvn** [comment]
>  I can't think of any easily repeatable ones either for testing that use case. I'd say it's best to put the delay in nonetheless for the sake of clean screenshots.

Using a duel with a quick forfeit for testing, achievement for the honorable medallion and then the duelicious achievement in rapid succession.

**Comment by kaytotes** [comment]
Been out of action for a couple days needing to get new glasses so only just looked at this now. This looks fine to me. Great pull request.

**Pull request merged**